File: mobile/app/loan-calculator.js
// Loan and amortization calculator. Shows the real monthly payment, the total
// interest, and, most importantly, the TRUE effective rate, so a low "add-on"
// quote cannot hide how much a loan really costs. All math is in lib/loan.js
// (pure and tested). This is an estimate from the numbers you enter, not a loan
// offer, and the screen says so.

import { useMemo, useState } from 'react';
import { Pressable, ScrollView, StyleSheet, Text, TextInput, View } from 'react-native';
import { SafeAreaView } from 'react-native-safe-area-context';
import { useRouter } from 'expo-router';
import { Ionicons } from '@expo/vector-icons';
import { spacing, radius, fontSize, fontWeight } from '../theme';
import { useTheme } from '../context/Theme';
import { formatMoney } from '../lib/format';
import { loanSummary } from '../lib/loan';

export default function LoanCalculator() {
  const { colors } = useTheme();
  const styles = useMemo(() => makeStyles(colors), [colors]);
  const router = useRouter();

  const [amount, setAmount] = useState('');
  const [term, setTerm] = useState('');
  const [termUnit, setTermUnit] = useState('months'); // 'months' | 'years'
  const [rate, setRate] = useState('');
  const [rateBasis, setRateBasis] = useState('monthly'); // 'monthly' | 'annual'
  const [method, setMethod] = useState('diminishing'); // 'diminishing' | 'addon'
  const [showSchedule, setShowSchedule] = useState(false);

  const parse = (s) => Number(String(s).replace(/[, ]/g, '')) || 0;
  const amountNum = parse(amount);
  const termNum = parse(term);
  const rateNum = parse(rate);
  const months = termUnit === 'years' ? Math.round(termNum * 12) : Math.round(termNum);

  const r = useMemo(
    () => loanSummary(amountNum, rateNum, months, { method, rateBasis }),
    [amountNum, rateNum, months, method, rateBasis]
  );
  const m = (n) => formatMoney(Math.round(n));
  const pct = (x) => (x * 100).toFixed(2) + '%';

  const ready = amountNum > 0 && months >= 1 && rateNum >= 0;
  const addon = method === 'addon';

  const Seg = ({ options, value, onChange }) => (
    <View style={styles.segment}>
      {options.map((o) => (
        <Pressable key={o.id} style={[styles.segBtn, value === o.id && styles.segBtnOn]} onPress={() => onChange(o.id)}>
          <Text style={[styles.segText, value === o.id && styles.segTextOn]}>{o.label}</Text>
        </Pressable>
      ))}
    </View>
  );

  return (
    <SafeAreaView style={styles.screen} edges={['top']}>
      <View style={styles.headerBar}>
        <Pressable onPress={() => router.back()} hitSlop={10} style={styles.back}>
          <Ionicons name="chevron-back" size={24} color={colors.text} />
        </Pressable>
        <Text style={styles.headerTitle}>Loan calculator</Text>
        <View style={{ width: 24 }} />
      </View>

      <ScrollView contentContainerStyle={styles.content} keyboardShouldPersistTaps="handled">
        <Text style={styles.intro}>
          See the real monthly payment and the true cost of a loan. If your lender quoted an add-on rate, this shows what it really works out to.
        </Text>

        <Text style={styles.fieldLabel}>Loan amount</Text>
        <View style={styles.inputWrap}>
          <Text style={styles.peso}>₱</Text>
          <TextInput style={styles.input} value={amount} onChangeText={setAmount} keyboardType="numeric" placeholder="e.g. 100,000" placeholderTextColor={colors.faint} autoFocus />
        </View>

        <View style={styles.twoCol}>
          <View style={styles.col}>
            <Text style={styles.fieldLabel}>Term</Text>
            <View style={styles.smallInputWrap}>
              <TextInput style={styles.smallInput} value={term} onChangeText={setTerm} keyboardType="numeric" placeholder="12" placeholderTextColor={colors.faint} />
            </View>
            <Seg options={[{ id: 'months', label: 'Months' }, { id: 'years', label: 'Years' }]} value={termUnit} onChange={setTermUnit} />
          </View>
          <View style={styles.col}>
            <Text style={styles.fieldLabel}>Interest rate</Text>
            <View style={styles.smallInputWrap}>
              <TextInput style={styles.smallInput} value={rate} onChangeText={setRate} keyboardType="numeric" placeholder="1.5" placeholderTextColor={colors.faint} />
              <Text style={styles.pctSign}>%</Text>
            </View>
            <Seg options={[{ id: 'monthly', label: 'Per month' }, { id: 'annual', label: 'Per year' }]} value={rateBasis} onChange={setRateBasis} />
          </View>
        </View>

        <Text style={[styles.fieldLabel, { marginTop: spacing.lg }]}>How the interest is charged</Text>
        <Seg options={[{ id: 'diminishing', label: 'Diminishing' }, { id: 'addon', label: 'Add-on' }]} value={method} onChange={setMethod} />
        <Text style={styles.methodHint}>
          {addon
            ? 'Add-on charges interest on the full amount for the whole term, even the part you have already paid back. Common in in-house and informal financing, and much costlier than it looks.'
            : 'Diminishing balance charges interest only on what you still owe. This is how banks and most formal lenders compute a loan.'}
        </Text>

        {ready ? (
          <>
            <View style={styles.card}>
              <Text style={styles.payLabel}>Monthly payment</Text>
              <Text style={styles.payValue}>{m(r.payment)}</Text>
              <View style={styles.splitRow}>
                <View style={styles.splitCol}>
                  <Text style={styles.splitLabel}>Total interest</Text>
                  <Text style={styles.splitValue}>{m(r.totalInterest)}</Text>
                </View>
                <View style={styles.splitCol}>
                  <Text style={styles.splitLabel}>Total to pay</Text>
                  <Text style={styles.splitValue}>{m(r.totalPaid)}</Text>
                </View>
              </View>
            </View>

            <View style={[styles.rateCard, addon && styles.rateCardWarn]}>
              <Text style={styles.rateKicker}>TRUE COST</Text>
              <View style={styles.rateRow}>
                <Text style={styles.rateLabel}>Effective interest per year</Text>
                <Text style={[styles.rateValue, addon && styles.rateValueWarn]}>{pct(r.effectiveAnnualRate)}</Text>
              </View>
              <Text style={styles.rateNote}>
                {addon
                  ? `Your ${pct(r.quotedMonthlyRate)} a month add-on really costs about ${pct(r.effectiveMonthlyRate)} a month, or ${pct(r.effectiveAnnualRate)} a year, once you account for paying interest on money you have already returned.`
                  : `This is the yearly rate with monthly compounding. On a diminishing balance the effective rate is close to the quoted rate.`}
              </Text>
            </View>

            <Pressable style={styles.scheduleToggle} onPress={() => setShowSchedule((v) => !v)}>
              <Text style={styles.scheduleToggleText}>{showSchedule ? 'Hide payment schedule' : 'Show payment schedule'}</Text>
              <Ionicons name={showSchedule ? 'chevron-up' : 'chevron-down'} size={16} color={colors.primary} />
            </Pressable>

            {showSchedule ? (
              <View style={styles.scheduleCard}>
                <View style={styles.schHead}>
                  <Text style={[styles.schCol, styles.schNum]}>#</Text>
                  <Text style={[styles.schCol, styles.schAmt]}>Interest</Text>
                  <Text style={[styles.schCol, styles.schAmt]}>Principal</Text>
                  <Text style={[styles.schCol, styles.schAmt]}>Balance</Text>
                </View>
                {r.schedule.map((row) => (
                  <View key={row.period} style={styles.schRow}>
                    <Text style={[styles.schCol, styles.schNum, styles.schMuted]}>{row.period}</Text>
                    <Text style={[styles.schCol, styles.schAmt, styles.schMuted]}>{m(row.interest)}</Text>
                    <Text style={[styles.schCol, styles.schAmt]}>{m(row.principal)}</Text>
                    <Text style={[styles.schCol, styles.schAmt, styles.schMuted]}>{m(row.balance)}</Text>
                  </View>
                ))}
              </View>
            ) : null}
          </>
        ) : (
          <Text style={styles.hint}>Enter the loan amount, term, and interest rate to see the payment and the true cost.</Text>
        )}

        <Text style={styles.disclaimer}>
          An estimate from the numbers you enter. Real loans add fees, insurance, and penalties, and a pre-termination charge can reduce the saving from paying early. This is a guide, not a loan offer. If a lender will not tell you the effective interest rate, that is a warning sign.
        </Text>
      </ScrollView>
    </SafeAreaView>
  );
}

function makeStyles(colors) {
  return StyleSheet.create({
    screen: { flex: 1, backgroundColor: colors.background },
    headerBar: { flexDirection: 'row', alignItems: 'center', justifyContent: 'space-between', paddingHorizontal: spacing.lg, paddingTop: spacing.md, paddingBottom: spacing.sm },
    back: { marginLeft: -4 },
    headerTitle: { color: colors.text, fontSize: fontSize.subtitle, fontWeight: fontWeight.bold },
    content: { padding: spacing.lg, paddingBottom: spacing.xxl },
    intro: { color: colors.muted, fontSize: fontSize.small, lineHeight: 19, marginBottom: spacing.lg },

    fieldLabel: { color: colors.muted, fontSize: fontSize.caption, marginBottom: spacing.sm, letterSpacing: 0.3 },
    inputWrap: { flexDirection: 'row', alignItems: 'center', backgroundColor: colors.card, borderColor: colors.border, borderWidth: 1, borderRadius: radius.md, paddingHorizontal: spacing.lg, marginBottom: spacing.lg },
    peso: { color: colors.muted, fontSize: fontSize.title, fontWeight: fontWeight.bold, marginRight: spacing.sm },
    input: { flex: 1, paddingVertical: spacing.md, color: colors.text, fontSize: fontSize.title, fontWeight: fontWeight.bold },

    twoCol: { flexDirection: 'row', gap: spacing.md },
    col: { flex: 1 },
    smallInputWrap: { flexDirection: 'row', alignItems: 'center', backgroundColor: colors.card, borderColor: colors.border, borderWidth: 1, borderRadius: radius.md, paddingHorizontal: spacing.md, marginBottom: spacing.sm },
    smallInput: { flex: 1, paddingVertical: spacing.md, color: colors.text, fontSize: fontSize.body, fontWeight: fontWeight.bold },
    pctSign: { color: colors.muted, fontSize: fontSize.body, fontWeight: fontWeight.bold, marginLeft: spacing.xs },

    segment: { flexDirection: 'row', backgroundColor: colors.card, borderColor: colors.border, borderWidth: 1, borderRadius: radius.md, padding: 3 },
    segBtn: { flex: 1, paddingVertical: spacing.sm, alignItems: 'center', borderRadius: radius.sm },
    segBtnOn: { backgroundColor: colors.primary },
    segText: { color: colors.muted, fontSize: fontSize.small, fontWeight: fontWeight.medium },
    segTextOn: { color: colors.background, fontWeight: fontWeight.bold },
    methodHint: { color: colors.faint, fontSize: fontSize.caption, lineHeight: 16, marginTop: spacing.sm },

    card: { backgroundColor: colors.card, borderColor: colors.border, borderWidth: 1, borderRadius: radius.lg, padding: spacing.lg, marginTop: spacing.xl },
    payLabel: { color: colors.muted, fontSize: fontSize.caption, letterSpacing: 0.3 },
    payValue: { color: colors.primary, fontSize: fontSize.huge, fontWeight: fontWeight.heavy, marginTop: 2 },
    splitRow: { flexDirection: 'row', borderTopColor: colors.border, borderTopWidth: StyleSheet.hairlineWidth, marginTop: spacing.md, paddingTop: spacing.md },
    splitCol: { flex: 1 },
    splitLabel: { color: colors.muted, fontSize: fontSize.caption },
    splitValue: { color: colors.text, fontSize: fontSize.body, fontWeight: fontWeight.bold, marginTop: 2 },

    rateCard: { backgroundColor: colors.card, borderColor: colors.border, borderWidth: 1, borderRadius: radius.lg, padding: spacing.lg, marginTop: spacing.md },
    rateCardWarn: { borderColor: colors.warning || colors.primary },
    rateKicker: { color: colors.muted, fontSize: fontSize.caption, fontWeight: fontWeight.bold, letterSpacing: 1.2 },
    rateRow: { flexDirection: 'row', justifyContent: 'space-between', alignItems: 'center', marginTop: spacing.sm },
    rateLabel: { color: colors.text, fontSize: fontSize.body, fontWeight: fontWeight.medium, flexShrink: 1, paddingRight: spacing.md },
    rateValue: { color: colors.text, fontSize: fontSize.title, fontWeight: fontWeight.heavy },
    rateValueWarn: { color: colors.warning || colors.primary },
    rateNote: { color: colors.textSecondary, fontSize: fontSize.small, lineHeight: 19, marginTop: spacing.sm },

    scheduleToggle: { flexDirection: 'row', alignItems: 'center', justifyContent: 'center', gap: spacing.xs, paddingVertical: spacing.md, marginTop: spacing.md },
    scheduleToggleText: { color: colors.primary, fontSize: fontSize.small, fontWeight: fontWeight.bold },
    scheduleCard: { backgroundColor: colors.card, borderColor: colors.border, borderWidth: 1, borderRadius: radius.lg, padding: spacing.md },
    schHead: { flexDirection: 'row', borderBottomColor: colors.border, borderBottomWidth: StyleSheet.hairlineWidth, paddingBottom: spacing.sm, marginBottom: spacing.xs },
    schRow: { flexDirection: 'row', paddingVertical: spacing.xs },
    schCol: { fontSize: fontSize.caption },
    schNum: { width: 28, color: colors.text },
    schAmt: { flex: 1, textAlign: 'right', color: colors.text, fontWeight: fontWeight.medium },
    schMuted: { color: colors.muted, fontWeight: fontWeight.regular },

    hint: { color: colors.muted, fontSize: fontSize.body, textAlign: 'center', marginTop: spacing.xl, marginBottom: spacing.xl },
    disclaimer: { color: colors.faint, fontSize: fontSize.caption, lineHeight: 17, marginTop: spacing.xl },
  });
}

File: flutter/lib/screens/loan_calculator.dart
// Loan calculator: the RN tool adapted to Flutter on the golden-ported
// loan math. Its one job is honesty: show the real monthly payment and the
// TRUE effective rate, so a lender's friendly add-on quote cannot hide an
// effective rate roughly double it. All estimates; real contracts add fees
// and penalties, and the copy says so.

import 'package:flutter/material.dart';

import '../money/loan.dart';
import '../theme.dart';
import '../typography.dart';
import 'overview.dart' show formatMoney;

class LoanCalculatorScreen extends StatefulWidget {
  const LoanCalculatorScreen({super.key});

  @override
  State<LoanCalculatorScreen> createState() => _LoanCalculatorScreenState();
}

class _LoanCalculatorScreenState extends State<LoanCalculatorScreen> {
  final amount = TextEditingController();
  final term = TextEditingController();
  final rate = TextEditingController();
  String termUnit = 'months';
  String rateBasis = 'monthly';
  String method = 'diminishing';
  bool showSchedule = false;

  @override
  void dispose() {
    amount.dispose();
    term.dispose();
    rate.dispose();
    super.dispose();
  }

  double _parse(String s) {
    final cleaned = s.replaceAll(RegExp(r'[, ]'), '');
    if (cleaned.isEmpty) return 0;
    final n = double.tryParse(cleaned);
    return (n == null || !n.isFinite) ? 0 : n;
  }

  String _m(num n) => formatMoney((n + 0.5).floorToDouble());

  String _pct(num x) => '${(x * 100).toStringAsFixed(2)}%';

  Widget _seg(
    List<(String, String)> options,
    String value,
    void Function(String) onChange,
  ) {
    return Wrap(
      spacing: 8,
      children: [
        for (final o in options)
          ChoiceChip(
            label: Text(o.$2),
            selected: value == o.$1,
            onSelected: (_) => setState(() => onChange(o.$1)),
            selectedColor: Barako.primary,
            backgroundColor: Barako.background,
            labelStyle: AppText.caption.w6.tint(
              value == o.$1 ? Barako.onPrimary : Barako.textSecondary,
            ),
          ),
      ],
    );
  }

  Widget _label(String text) => Padding(
    padding: const EdgeInsets.only(top: 14, bottom: 6),
    child: Text(text, style: AppText.caption.w7),
  );

  @override
  Widget build(BuildContext context) {
    final amountNum = _parse(amount.text);
    final termNum = _parse(term.text);
    final rateNum = _parse(rate.text);
    // Clamp in DOUBLE space before toInt: a pasted 2e307 in years mode
    // multiplies to Infinity, and Infinity.toInt() throws where the RN
    // screen's Math.min clamp survives. Same result, no crash.
    final rawMonthsD = termUnit == 'years'
        ? (termNum * 12 + 0.5).floorToDouble()
        : (termNum + 0.5).floorToDouble();
    final termClamped = rawMonthsD > maxMonths;
    // clamp handles both infinities: +Infinity lands on the cap like RN's
    // Math.min, -Infinity lands on zero so a pasted negative monster shows
    // only the negative warning, never a result card beneath it.
    final months = rawMonthsD.clamp(0, maxMonths.toDouble()).toInt();

    final ready = amountNum > 0 && months >= 1 && rateNum >= 0;
    final badInput = amountNum < 0 || termNum < 0 || rateNum < 0;
    final needTerm =
        !ready && !badInput && amountNum > 0 && rateNum >= 0 && rawMonthsD < 1;
    final addon = method == 'addon';

    final r = ready
        ? loanSummary(
            amountNum,
            rateNum,
            months,
            method: method,
            rateBasis: rateBasis,
          )
        : null;
    // The payoff nugget uses the EXACT rate, not the display-rounded
    // quotedMonthlyRate, or its balance would contradict the schedule row
    // shown on the same screen (bank officer finding).
    final exactMonthlyRate = rateBasis == 'annual'
        ? rateNum / 100 / 12
        : rateNum / 100;
    final payoff = r != null
        ? payoffSaving(amountNum, exactMonthlyRate, months, months ~/ 2)
        : null;

    return Scaffold(
      appBar: AppBar(
        backgroundColor: Barako.background,
        foregroundColor: Barako.text,
        title: Text(
          'Loan calculator',
          style: TextStyle(color: Barako.text, fontWeight: FontWeight.w800),
        ),
      ),
      body: SafeArea(
        child: ListView(
          padding: const EdgeInsets.fromLTRB(20, 8, 20, 32),
          children: [
            Text(
              'See the real monthly payment and the true cost of a loan. If your lender quoted an add-on rate, this shows what it really works out to.',
              style: AppText.small.tint(Barako.muted).copyWith(height: 1.4),
            ),
            _label('LOAN AMOUNT'),
            TextField(
              controller: amount,
              onChanged: (_) => setState(() {}),
              keyboardType: const TextInputType.numberWithOptions(
                decimal: true,
              ),
              autofocus: true,
              style: TextStyle(color: Barako.text),
              decoration: InputDecoration(
                prefixText: '₱ ',
                hintText: 'e.g. 100,000',
              ),
            ),
            _label('TERM'),
            Row(
              children: [
                SizedBox(
                  width: 110,
                  child: TextField(
                    controller: term,
                    onChanged: (_) => setState(() {}),
                    keyboardType: TextInputType.number,
                    style: TextStyle(color: Barako.text),
                    decoration: InputDecoration(hintText: 'e.g. 12'),
                  ),
                ),
                const SizedBox(width: 12),
                _seg(
                  const [('months', 'Months'), ('years', 'Years')],
                  termUnit,
                  (v) => termUnit = v,
                ),
              ],
            ),
            _label('INTEREST RATE'),
            Row(
              children: [
                SizedBox(
                  width: 110,
                  child: TextField(
                    controller: rate,
                    onChanged: (_) => setState(() {}),
                    keyboardType: const TextInputType.numberWithOptions(
                      decimal: true,
                    ),
                    style: TextStyle(color: Barako.text),
                    decoration: InputDecoration(
                      hintText: 'e.g. 1.5',
                      suffixText: '%',
                    ),
                  ),
                ),
                const SizedBox(width: 12),
                _seg(
                  const [('monthly', 'Per month'), ('annual', 'Per year')],
                  rateBasis,
                  (v) => rateBasis = v,
                ),
              ],
            ),
            _label('HOW THE INTEREST IS CHARGED'),
            _seg(
              const [('diminishing', 'Diminishing'), ('addon', 'Add-on')],
              method,
              (v) => method = v,
            ),
            const SizedBox(height: 6),
            Text(
              addon
                  ? 'Add-on charges interest on the ORIGINAL amount for the whole term, common in in-house and informal financing. It looks cheap and is not.'
                  : 'Diminishing charges interest only on what you still owe, the way banks amortize.',
              style: AppText.caption.copyWith(height: 1.4),
            ),
            if (termClamped)
              Padding(
                padding: const EdgeInsets.only(top: 8),
                child: Text(
                  'Term capped at $maxMonths months, no consumer loan runs a century.',
                  style: AppText.caption.tint(Barako.warning),
                ),
              ),
            if (badInput)
              Padding(
                padding: const EdgeInsets.only(top: 12),
                child: Text(
                  'Amounts and rates cannot be negative.',
                  style: AppText.small.tint(Barako.warning),
                ),
              )
            else if (needTerm)
              Padding(
                padding: const EdgeInsets.only(top: 12),
                child: Text(
                  'Enter the term and the numbers appear.',
                  style: AppText.small.tint(Barako.muted),
                ),
              )
            else if (!ready)
              Padding(
                padding: const EdgeInsets.only(top: 12),
                child: Text(
                  'Enter the loan amount, term, and interest rate and the numbers appear.',
                  style: AppText.small.tint(Barako.muted),
                ),
              ),
            if (r != null) ...[
              const SizedBox(height: 16),
              Card(
                child: Padding(
                  padding: const EdgeInsets.all(16),
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text(
                        'MONTHLY PAYMENT',
                        style: TextStyle(
                          color: Barako.muted,
                          fontSize: 11,
                          fontWeight: FontWeight.w700,
                          letterSpacing: 2,
                        ),
                      ),
                      const SizedBox(height: 4),
                      Text(
                        _m(r['payment'] as double),
                        style: AppText.amountLg.w7,
                      ),
                      const SizedBox(height: 8),
                      Row(
                        children: [
                          Expanded(
                            child: Text('Total interest', style: AppText.small),
                          ),
                          Text(
                            _m(r['totalInterest'] as double),
                            style: AppText.small.w6.tabular,
                          ),
                        ],
                      ),
                      Row(
                        children: [
                          Expanded(
                            child: Text('Total to pay', style: AppText.small),
                          ),
                          Text(
                            _m(r['totalPaid'] as double),
                            style: AppText.small.w6.tabular,
                          ),
                        ],
                      ),
                    ],
                  ),
                ),
              ),
              const SizedBox(height: 12),
              Card(
                color: addon ? null : Barako.card,
                child: Padding(
                  padding: const EdgeInsets.all(16),
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text(
                        'TRUE COST',
                        style: TextStyle(
                          color: addon ? Barako.warning : Barako.muted,
                          fontSize: 11,
                          fontWeight: FontWeight.w700,
                          letterSpacing: 2,
                        ),
                      ),
                      const SizedBox(height: 6),
                      Row(
                        children: [
                          Expanded(
                            child: Text(
                              'Quoted rate per year',
                              style: AppText.small,
                            ),
                          ),
                          Text(
                            _pct(r['nominalAnnualRate'] as double),
                            style: AppText.small,
                          ),
                        ],
                      ),
                      Row(
                        children: [
                          Expanded(
                            child: Text(
                              'Effective interest per year',
                              style: AppText.smallStrong,
                            ),
                          ),
                          Text(
                            _pct(r['effectiveAnnualRate'] as double),
                            style: AppText.body.w8.tint(
                              addon ? Barako.warning : Barako.text,
                            ),
                          ),
                        ],
                      ),
                      const SizedBox(height: 6),
                      Text(
                        addon
                            ? 'Your ${rateBasis == 'annual' ? '${_pct(r['nominalAnnualRate'] as double)} a year' : '${_pct(r['quotedMonthlyRate'] as double)} a month'} add-on really works out to about ${_pct(r['effectiveAnnualRate'] as double)} a year, once you account for paying interest on money you have already returned.'
                            : 'For a diminishing loan the quoted and effective rates differ only because of monthly compounding. At normal bank rates they stay close.',
                        style: AppText.caption.copyWith(height: 1.4),
                      ),
                    ],
                  ),
                ),
              ),
              if (payoff != null &&
                  months >= 2 &&
                  (payoff['interestSaved'] as double) > 0 &&
                  method == 'diminishing') ...[
                const SizedBox(height: 12),
                Card(
                  child: Padding(
                    padding: const EdgeInsets.all(16),
                    child: Text(
                      'Make your first ${months ~/ 2} payments, then pay the remaining ${_m(payoff['balanceCleared'] as double)} in one go, and you skip about ${_m(payoff['interestSaved'] as double)} of the remaining interest, if there is no pre-termination fee.',
                      style: AppText.small.copyWith(height: 1.4),
                    ),
                  ),
                ),
              ],
              const SizedBox(height: 12),
              TextButton(
                onPressed: () => setState(() => showSchedule = !showSchedule),
                child: Text(
                  showSchedule
                      ? 'Hide the month by month schedule'
                      : 'Show the month by month schedule',
                ),
              ),
              if (showSchedule)
                Card(
                  child: Padding(
                    padding: const EdgeInsets.all(12),
                    child: Column(
                      children: [
                        for (final row
                            in (r['schedule'] as List)
                                .cast<Map<String, dynamic>>())
                          Padding(
                            padding: const EdgeInsets.symmetric(vertical: 3),
                            child: Row(
                              children: [
                                SizedBox(
                                  width: 34,
                                  child: Text(
                                    '${row['period']}',
                                    style: AppText.caption,
                                  ),
                                ),
                                Expanded(
                                  child: Text(
                                    '${_m(row['interest'] as double)} interest · ${_m(row['principal'] as double)} principal',
                                    style: AppText.caption.tint(
                                      Barako.textSecondary,
                                    ),
                                  ),
                                ),
                                Text(
                                  _m(row['balance'] as double),
                                  style: AppText.caption.w6
                                      .tint(Barako.textSecondary)
                                      .tabular,
                                ),
                              ],
                            ),
                          ),
                      ],
                    ),
                  ),
                ),
              const SizedBox(height: 12),
              Text(
                'This is an estimate from the numbers you typed, not a loan offer. Real contracts add fees, insurance, and penalties, and a pre-termination charge can reduce the saving from paying early. Read the disclosure statement before signing. If a lender will not tell you the effective interest rate, that is a warning sign.',
                style: AppText.micro.w4
                    .tint(Barako.faint)
                    .copyWith(height: 1.4),
              ),
            ],
          ],
        ),
      ),
    );
  }
}
